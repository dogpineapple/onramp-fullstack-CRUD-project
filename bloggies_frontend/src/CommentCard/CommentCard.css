.CommentCard {
  width: 40rem;
  max-width: fit-content;
}