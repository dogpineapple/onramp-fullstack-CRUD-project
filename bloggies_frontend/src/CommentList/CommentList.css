.CommentList {
  margin: 2rem auto;
}

.CommentsList-no-comments {
  margin: 1rem 1rem;
}