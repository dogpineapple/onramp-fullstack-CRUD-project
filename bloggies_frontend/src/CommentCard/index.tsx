import React from "react";
import { Card } from "react-bootstrap";
import CommentReplyCard from "../CommentReplyCard";
import { Comment } from "../custom";
import "./CommentCard.css";

interface IProp {
  comment: Comment
}

function CommentCard({ comment }:IProp) {
  return (
    <div className="CommentCard text-left">
      <Card>
        <Card.Body>
          <Card.Text>{comment.body}</Card.Text>
          <Card.Text className="text-muted">
            <span className="App-author">{comment.author_name}</span> commented {new Date(comment.created_at).toString()}
          </Card.Text>
          { comment.reply_count > 0 && <CommentReplyCard replyCount={comment.reply_count} commentId={comment.id}/>}
        </Card.Body>
      </Card>
    </div>
  );
};

export default CommentCard;