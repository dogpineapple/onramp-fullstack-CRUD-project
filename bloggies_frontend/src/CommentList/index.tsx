import React from "react";
import CommentCard from "../CommentCard";
import { Comment } from "../custom";
import CommentForm from "../Forms/CommentForm";
import "./CommentList.css";

interface IProp {
  comments: Array<Comment>,
  postId: number,
  postComment: Function
}

/**
 * `CommentList` renders `CommentCard` and a `CommentForm`.
 * It displays all comments that are NOT replies (ex: comment.is_reply === false )
 */
function CommentList({ comments, postId, postComment }: IProp) {

  /** handler fxn for posting a comment (to be invoked in CommentForm (for non-replies) and CommentCard (for replies))*/
  const handlePostComment = (postId: number, commentId: number | undefined, isReply: boolean, comment: string) => {
    postComment(postId, commentId, isReply, comment);
  }

  const nonReplyComments = comments.filter(c => {
    return !c.is_reply;
  })

  return (
    <div className="CommentList text-left">
      <h3>Comments</h3>
      {nonReplyComments.length > 0 ?
        nonReplyComments.map(c => {
          return (
            <CommentCard key={c.id} comment={c} handlePostReply={handlePostComment}/>
          );
        })
        : <div className="CommentsList-no-comments">No comments yet. Be the first!</div>}
      <CommentForm postId={postId} commentId={undefined} isReply={false} handlePostComment={handlePostComment}/>
    </div>
  );
};

export default CommentList;