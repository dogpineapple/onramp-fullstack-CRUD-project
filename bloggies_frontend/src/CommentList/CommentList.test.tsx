import React, { Component } from 'react';
import Enzyme, { mount, ReactWrapper, shallow } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import CommentList from './index';
import { MOCK_COMMENTS, MOCK_STORE, MOCK_POST } from '../jest.mock';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import CommentCard from '../CommentCard';

Enzyme.configure({ adapter: new Adapter() });

const mockStore = configureStore();

const mockPostComment = jest.fn(() => "Posted");

describe('CommentList', () => {
  let wrapper: ReactWrapper;

  beforeEach(() => {
    wrapper = mount(shallow<Component>(
      <Provider store={mockStore(MOCK_STORE)}>
        <BrowserRouter>
          <CommentList comments={MOCK_COMMENTS} postId={MOCK_POST.id} postComment={mockPostComment} />
        </BrowserRouter>
      </Provider>
    ).get(0));
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should display comments', () => {
    const commentCards = wrapper.find(CommentCard);
    const firstComment = wrapper.containsMatchingElement(
      <p className="sc-bdfBwQ dhWHrV card-text">
        first test comment
      </p>
    );

    expect(commentCards).toHaveLength(2);
    expect(firstComment).toBeTruthy();
  });

  it('should display "No comments yet" with no comments', () => {
    wrapper = mount(shallow<Component>(
      <Provider store={mockStore(MOCK_STORE)}>
        <BrowserRouter>
          <CommentList comments={[]} postId={MOCK_POST.id} postComment={mockPostComment} />
        </BrowserRouter>
      </Provider>
    ).get(0));

    const elementFound = wrapper.containsMatchingElement(
      <div className="CommentsList-no-comments">No comments yet. Be the first!</div>
    );

    expect(elementFound).toBeTruthy();
  });
});