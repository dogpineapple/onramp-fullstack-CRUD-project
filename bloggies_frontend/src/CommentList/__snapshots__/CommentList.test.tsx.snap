// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentList should match snapshot 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <CommentList
      comments={
        Array [
          Object {
            "author_id": 1,
            "author_name": "test user",
            "body": "first test comment",
            "created_at": "1/1/2021",
            "id": 1,
            "is_reply": false,
            "post_id": 1,
            "reply_count": "1",
          },
          Object {
            "author_id": 1,
            "author_name": "test user",
            "body": "second test comment",
            "created_at": "1/1/2021",
            "id": 2,
            "is_reply": false,
            "post_id": 1,
            "reply_count": "1",
          },
        ]
      }
      postComment={[MockFunction]}
      postId={123}
    >
      <div
        className="CommentList text-left"
      >
        <h3>
          Comments
        </h3>
        <CommentCard
          comment={
            Object {
              "author_id": 1,
              "author_name": "test user",
              "body": "first test comment",
              "created_at": "1/1/2021",
              "id": 1,
              "is_reply": false,
              "post_id": 1,
              "reply_count": "1",
            }
          }
          handlePostReply={[Function]}
          key="1"
        >
          <div
            className="CommentCard text-left"
          >
            <Card
              body={false}
            >
              <div
                className="card"
              >
                <CardBody>
                  <div
                    className="card-body"
                  >
                    <CardText>
                      <p
                        className="card-text"
                      >
                        first test comment
                      </p>
                    </CardText>
                    <CardText
                      className="text-muted d-flex justify-content-between"
                    >
                      <p
                        className="text-muted d-flex justify-content-between card-text"
                      >
                        <span>
                          <span
                            className="App-author"
                          >
                            test user
                          </span>
                           commented 
                          a month ago
                        </span>
                        <Button
                          active={false}
                          disabled={false}
                          onClick={[Function]}
                          variant="primary"
                        >
                          <button
                            className="btn btn-primary"
                            disabled={false}
                            onClick={[Function]}
                            type="button"
                          >
                            Reply
                          </button>
                        </Button>
                      </p>
                    </CardText>
                    <CommentReplyAccord
                      commentId={1}
                      key="1"
                      replyCount="1"
                    >
                      <div
                        className="CommentReplyAccord mt-3"
                      >
                        <Accordion
                          defaultActiveKey="0"
                        >
                          <div
                            className="accordion"
                          >
                            <Card
                              body={false}
                            >
                              <div
                                className="card"
                              >
                                <CardHeader>
                                  <div
                                    className="card-header"
                                  >
                                    <ForwardRef
                                      as={
                                        Object {
                                          "$$typeof": Symbol(react.forward_ref),
                                          "defaultProps": Object {
                                            "active": false,
                                            "disabled": false,
                                            "variant": "primary",
                                          },
                                          "render": [Function],
                                        }
                                      }
                                      eventKey="1"
                                      onClick={[Function]}
                                      variant="link"
                                    >
                                      <Button
                                        active={false}
                                        disabled={false}
                                        onClick={[Function]}
                                        variant="link"
                                      >
                                        <button
                                          className="btn btn-link"
                                          disabled={false}
                                          onClick={[Function]}
                                          type="button"
                                        >
                                          View 
                                          1
                                           
                                          reply
                                        </button>
                                      </Button>
                                    </ForwardRef>
                                  </div>
                                </CardHeader>
                                <AccordionCollapse
                                  eventKey="1"
                                >
                                  <ForwardRef
                                    appear={false}
                                    getDimensionValue={[Function]}
                                    in={false}
                                    mountOnEnter={false}
                                    timeout={300}
                                    unmountOnExit={false}
                                  >
                                    <Transition
                                      addEndListener={[Function]}
                                      appear={false}
                                      aria-expanded={null}
                                      enter={true}
                                      exit={true}
                                      in={false}
                                      mountOnEnter={false}
                                      onEnter={[Function]}
                                      onEntered={[Function]}
                                      onEntering={[Function]}
                                      onExit={[Function]}
                                      onExited={[Function]}
                                      onExiting={[Function]}
                                      timeout={300}
                                      unmountOnExit={false}
                                    >
                                      <div
                                        aria-expanded={null}
                                        className="collapse"
                                      >
                                        <CardBody>
                                          <div
                                            className="card-body"
                                          >
                                            Fetching replies..
                                          </div>
                                        </CardBody>
                                      </div>
                                    </Transition>
                                  </ForwardRef>
                                </AccordionCollapse>
                              </div>
                            </Card>
                          </div>
                        </Accordion>
                      </div>
                    </CommentReplyAccord>
                  </div>
                </CardBody>
              </div>
            </Card>
          </div>
        </CommentCard>
        <CommentCard
          comment={
            Object {
              "author_id": 1,
              "author_name": "test user",
              "body": "second test comment",
              "created_at": "1/1/2021",
              "id": 2,
              "is_reply": false,
              "post_id": 1,
              "reply_count": "1",
            }
          }
          handlePostReply={[Function]}
          key="2"
        >
          <div
            className="CommentCard text-left"
          >
            <Card
              body={false}
            >
              <div
                className="card"
              >
                <CardBody>
                  <div
                    className="card-body"
                  >
                    <CardText>
                      <p
                        className="card-text"
                      >
                        second test comment
                      </p>
                    </CardText>
                    <CardText
                      className="text-muted d-flex justify-content-between"
                    >
                      <p
                        className="text-muted d-flex justify-content-between card-text"
                      >
                        <span>
                          <span
                            className="App-author"
                          >
                            test user
                          </span>
                           commented 
                          a month ago
                        </span>
                        <Button
                          active={false}
                          disabled={false}
                          onClick={[Function]}
                          variant="primary"
                        >
                          <button
                            className="btn btn-primary"
                            disabled={false}
                            onClick={[Function]}
                            type="button"
                          >
                            Reply
                          </button>
                        </Button>
                      </p>
                    </CardText>
                    <CommentReplyAccord
                      commentId={2}
                      key="1"
                      replyCount="1"
                    >
                      <div
                        className="CommentReplyAccord mt-3"
                      >
                        <Accordion
                          defaultActiveKey="0"
                        >
                          <div
                            className="accordion"
                          >
                            <Card
                              body={false}
                            >
                              <div
                                className="card"
                              >
                                <CardHeader>
                                  <div
                                    className="card-header"
                                  >
                                    <ForwardRef
                                      as={
                                        Object {
                                          "$$typeof": Symbol(react.forward_ref),
                                          "defaultProps": Object {
                                            "active": false,
                                            "disabled": false,
                                            "variant": "primary",
                                          },
                                          "render": [Function],
                                        }
                                      }
                                      eventKey="2"
                                      onClick={[Function]}
                                      variant="link"
                                    >
                                      <Button
                                        active={false}
                                        disabled={false}
                                        onClick={[Function]}
                                        variant="link"
                                      >
                                        <button
                                          className="btn btn-link"
                                          disabled={false}
                                          onClick={[Function]}
                                          type="button"
                                        >
                                          View 
                                          1
                                           
                                          reply
                                        </button>
                                      </Button>
                                    </ForwardRef>
                                  </div>
                                </CardHeader>
                                <AccordionCollapse
                                  eventKey="2"
                                >
                                  <ForwardRef
                                    appear={false}
                                    getDimensionValue={[Function]}
                                    in={false}
                                    mountOnEnter={false}
                                    timeout={300}
                                    unmountOnExit={false}
                                  >
                                    <Transition
                                      addEndListener={[Function]}
                                      appear={false}
                                      aria-expanded={null}
                                      enter={true}
                                      exit={true}
                                      in={false}
                                      mountOnEnter={false}
                                      onEnter={[Function]}
                                      onEntered={[Function]}
                                      onEntering={[Function]}
                                      onExit={[Function]}
                                      onExited={[Function]}
                                      onExiting={[Function]}
                                      timeout={300}
                                      unmountOnExit={false}
                                    >
                                      <div
                                        aria-expanded={null}
                                        className="collapse"
                                      >
                                        <CardBody>
                                          <div
                                            className="card-body"
                                          >
                                            Fetching replies..
                                          </div>
                                        </CardBody>
                                      </div>
                                    </Transition>
                                  </ForwardRef>
                                </AccordionCollapse>
                              </div>
                            </Card>
                          </div>
                        </Accordion>
                      </div>
                    </CommentReplyAccord>
                  </div>
                </CardBody>
              </div>
            </Card>
          </div>
        </CommentCard>
        <CommentForm
          handlePostComment={[Function]}
          isReply={false}
          postId={123}
        >
          <div
            className="CommentForm text-right"
          >
            <Form
              inline={false}
              onSubmit={[Function]}
            >
              <form
                className=""
                onSubmit={[Function]}
              >
                <FormControl
                  as="textarea"
                  name="comment"
                  onChange={[Function]}
                  placeholder="Enter comment"
                  required={true}
                  rows={3}
                  value=""
                >
                  <textarea
                    className="form-control"
                    name="comment"
                    onChange={[Function]}
                    placeholder="Enter comment"
                    required={true}
                    rows={3}
                    value=""
                  />
                </FormControl>
                <Button
                  active={false}
                  className="mt-1"
                  disabled={false}
                  type="submit"
                  variant="success"
                >
                  <button
                    className="mt-1 btn btn-success"
                    disabled={false}
                    type="submit"
                  >
                    Comment
                  </button>
                </Button>
              </form>
            </Form>
          </div>
        </CommentForm>
      </div>
    </CommentList>
  </Router>
</BrowserRouter>
`;
